### PR TITLE
Tensorboard plugin

### PR DESCRIPTION
**Pull request by ssheorey**
Summary writing support for all models (except KPConv) for both PyTorch and TF.
Address pylint warnings / errors in pipeline files (TF + PyTorch)
### WIP: 

- [x] Testing for all combinations (24). For all dataset split (train, test), framework (tf, torch) and models (6).
      Rename `.txt`->`.sh`, update dataset paths and run.
      [sum3d_tests.txt](https://github.com/isl-org/Open3D-ML/files/7458761/sum3d_tests.txt)
- [x] Order of label names for ScanNet

---
This change is [Reviewable](https://reviewable.io/reviews/isl-org/open3d-ml/405)

### CODE REVIEW

File: ml3d/torch/pipelines/semantic_segmentation.py [hunk 1]
```diff
@@ -436,19 +433,18 @@ def run_train(self):
                     self.metric_val.update(predict_scores, gt_labels)
 
                     self.valid_losses.append(loss.cpu().item())
+                    # Save only for the first batch
+                    if 'valid' in record_summary and step == 0:
```
[the remark this answers is not in the document]
Comment (ssheorey) [reply ml3d/torch/pipelines/semantic_segmentation.py:437]: Let's standardize on one word for this  -- different words in different places for the same functionality creates user confusion. We have different words currently. We should remove them in a future PR. Similar for `train` (not `training`) and `test` (not `testing`).

File: ml3d/torch/pipelines/semantic_segmentation.py [hunk 2]
```diff
@@ -380,6 +372,7 @@ def run_train(self):
         writer = SummaryWriter(self.tensorboard_dir)
         self.save_config(writer)
         log.info("Writing summary in {}.".format(self.tensorboard_dir))
+        record_summary = cfg.get('summary').get('record_for', [])
```
[the remark this answers is not in the document]
Comment (ssheorey) [reply ml3d/torch/pipelines/semantic_segmentation.py:375]: Added `self.cfg.setdefault('summary', {})` in `base_pipeline.py`  `__init__()` so that there is an empty dict if not provided

**Review by benjaminum — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/isl-org/open3d-ml/405)* status: 0 of 35 files reviewed, 5 unresolved discussions (waiting on @sanskar107 and @ssheorey)

___
*[ml3d/tf/pipelines/semantic_segmentation.py, line 317 at r10](https://reviewable.io/reviews/isl-org/open3d-ml/405#-Mo8NVR0E8nzzD-MZ77H:-Mo8NVR0E8nzzD-MZ77I:b-j29yen) ([raw file](https://github.com/isl-org/open3d-ml/blob/772e8f174165237e6733a1b72e44a744a8a924e5/ml3d/tf/pipelines/semantic_segmentation.py#L317)):*
> ```Python
>                 self.save_ckpt(epoch)
> 
>     def get_3d_summary(self, results, input_data, epoch, save_gt=True):
> ```

just a question. Would it be possible to unify the tf and torch code by converting early to numpy? Would it make sense to use numpy?
___
*[ml3d/vis/boundingbox.py, line 83 at r10](https://reviewable.io/reviews/isl-org/open3d-ml/405#-Mo8KJRh5ZGqZivMmOvD:-Mo8KJRh5ZGqZivMmOvE:b-j7ekhc) ([raw file](https://github.com/isl-org/open3d-ml/blob/772e8f174165237e6733a1b72e44a744a8a924e5/ml3d/vis/boundingbox.py#L83)):*
> ```Python
> 
>     @staticmethod
>     def create_lines(boxes, lut=None, out="lineset"):
> ```

`out -> out_format or format`. `out` could be mistaken for an output variable like it is used in many numpy functions.
___
*[ml3d/vis/boundingbox.py, line 99 at r10](https://reviewable.io/reviews/isl-org/open3d-ml/405#-Mo8MNsiBhZQpBjDl461:-Mo8MNsiBhZQpBjDl462:bt0ryb9) ([raw file](https://github.com/isl-org/open3d-ml/blob/772e8f174165237e6733a1b72e44a744a8a924e5/ml3d/vis/boundingbox.py#L99)):*
> ```Python
>                 "line_indices", "line_colors", "bbox_labels",
>                 "bbox_confidences") if out == "dict".
>         """
> ```

check output format here to provide a better error message.

**Review by ssheorey — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/isl-org/open3d-ml/405)* status: 0 of 35 files reviewed, 5 unresolved discussions (waiting on @benjaminum and @sanskar107)

___
*[ml3d/tf/pipelines/semantic_segmentation.py, line 317 at r10](https://reviewable.io/reviews/isl-org/open3d-ml/405#-Mo8NVR0E8nzzD-MZ77H:-MoBU0NB3LGtoPyvnKNE:b7dh51u) ([raw file](https://github.com/isl-org/open3d-ml/blob/772e8f174165237e6733a1b72e44a744a8a924e5/ml3d/tf/pipelines/semantic_segmentation.py#L317)):*
<details><summary><i>Previously, benjaminum (Benjamin Ummenhofer) wrote…</i></summary><blockquote>

just a question. Would it be possible to unify the tf and torch code by converting early to numpy? Would it make sense to use numpy?
</blockquote></details>

The models themselves have different representations for data, even in TF vs Torch. I don't think we'll be able to share code this way. The underlying `add_3d()` function works with any of TF, Torch, Numpy and Open3D tensors and converts everything to Open3D tensors for writing.
___
*[ml3d/vis/boundingbox.py, line 83 at r10](https://reviewable.io/reviews/isl-org/open3d-ml/405#-Mo8KJRh5ZGqZivMmOvD:-MoBWDP00edHf-YaT_jv:bb40hdc) ([raw file](https://github.com/isl-org/open3d-ml/blob/772e8f174165237e6733a1b72e44a744a8a924e5/ml3d/vis/boundingbox.py#L83)):*
<details><summary><i>Previously, benjaminum (Benjamin Ummenhofer) wrote…</i></summary><blockquote>

`out -> out_format or format`. `out` could be mistaken for an output variable like it is used in many numpy functions.
</blockquote></details>

Done. This needs a corresponding change to `summary.py` in the main repo. It will be in the `ssheorey/tb-plugin-docs` branch.
___
*[ml3d/vis/boundingbox.py, line 99 at r10](https://reviewable.io/reviews/isl-org/open3d-ml/405#-Mo8MNsiBhZQpBjDl461:-MoBWCfB0dO0DwL8i0SQ:b-896fix) ([raw file](https://github.com/isl-org/open3d-ml/blob/772e8f174165237e6733a1b72e44a744a8a924e5/ml3d/vis/boundingbox.py#L99)):*
<details><summary><i>Previously, benjaminum (Benjamin Ummenhofer) wrote…</i></summary><blockquote>

check output format here to provide a better error message.
</blockquote></details>

Done.

**Review by benjaminum — COMMENTED** [review]
from my side this looks good

> *[Reviewable](https://reviewable.io/reviews/isl-org/open3d-ml/405#-:-MoXlyJvANXAWg4ZpJ7M:b1r18kl)* status: 0 of 39 files reviewed, 5 unresolved discussions (waiting on @benjaminum and @sanskar107)

**Review by sanskar107 — APPROVED** [review]
